cohortextractor/dsl.py

```diff
@@ -0,0 +1,168 @@
+
+
+class EventFrame:
+    """Represents a collection of records, with multiple rows per patient.
+
+    Either an EventTable, or the result of filtering an EventTable.
+    """
+
+    def __init__(self, qm_table: QMBaseTable):
+        self.qm_table = qm_table
+
+    def filter(self, column: str, **kwargs: str) -> EventFrame:  # noqa: A003
+        """Return a new EventFrame with given filter."""
+
+        return EventFrame(self.qm_table.filter(column, **kwargs))
+
+    def sort_by(self, *columns: str) -> SortedEventFrame:
+        """Return a SortedEventFrame with given sort column."""
+
+        return SortedEventFrame(self.qm_table, *columns)
+
+    def count_for_patient(self) -> PatientSeries:
+        """Return a PatientSeries with count_for_patient of matching events per patient."""
+
+        return PatientSeries(self.qm_table.count())
+
+    def exists_for_patient(self) -> PatientSeries:
+        """Return a PatientSeries indicating whether each patient has a matching event."""
+
+        return PatientSeries(self.qm_table.exists())
+
+
+class SortedEventFrame:
+    """Represents an EventFrame that has been sorted."""
+
+    def __init__(self, qm_table: QMBaseTable, *sort_columns: str):
+        self.qm_table = qm_table
+        self.sort_columns = sort_columns
+
+    def first_for_patient(self) -> PatientFrame:
+        """Return a PatientFrame with the first event for each patient."""
+
+        return PatientFrame(row=self.qm_table.first_by(*self.sort_columns))
+
+    def last_for_patient(self) -> PatientFrame:
+        """Return a PatientFrame with the last event for each patient."""
+
+        return PatientFrame(row=self.qm_table.last_by(*self.sort_columns))
+
+
+class PatientFrame:
+    """Represents a collection of records, with one row per patient.
+
+    Either a PatientTable, or the result of filtering a PatientTable.
+    """
+
+    def __init__(self, row: Row):
+        self.row = row
+
+    def filter(self, column: str, **kwargs: str) -> PatientFrame:  # noqa: A003
+        """Return a new PatientFrame with given filter."""
+        # TODO
+
+    def select_column(self, column: str) -> PatientSeries:
+        """Return a PatientSeries containing given column."""
+
+        return PatientSeries(self.row.get(column))
+
+
+class PatientSeries:
+    """Represents a column indexed by patient.
+
+    Can be used as a variable in a Cohort, or as an input when computing another
+    variable.
+    """
+
+    def __init__(self, value: Value):
+        self.value = value
+
+
+class BaseTable:
+    """A base class for database tables."""
+
+    def __init__(self, name: str):
+        self.name = name
+        self.qm_table: QMBaseTable = QMTable(name)
+
+
+class EventTable(BaseTable, EventFrame):
+    """A base class for database tables with multiple rows per patient."""
+
+
+class PatientTable(BaseTable, PatientFrame):
+    """A base class for database tables with one row per patient."""
```

Comment:
I'm wary of multiple inheritance here.  A developer needs to know which order the `__init__`s are resolved in to work out which of those methods will be executed (as neither is calling `super()`).

I think at least BaseTable is a good candidate for a Protocol once the team understands them better.

Nothing to do for this PR though, just wanted to flag the potential oddities around MRO resolution!